File: apps/api/internal/adapter/driven/mongodb/invoice_repository.go
package mongodb

import (
	"context"
	"time"

	"github.com/neurodynecorp/api/internal/domain/entity"
	"github.com/neurodynecorp/api/internal/domain/port"
	"github.com/neurodynecorp/api/pkg/metrics"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

type InvoiceRepository struct {
	collection *mongo.Collection
	metrics    *metrics.Metrics
}

func NewInvoiceRepository(client *Client, m *metrics.Metrics) *InvoiceRepository {
	return &InvoiceRepository{
		collection: client.Collection("invoices"),
		metrics:    m,
	}
}

func (r *InvoiceRepository) Create(ctx context.Context, invoice *entity.Invoice) error {
	start := time.Now()
	defer func() {
		r.metrics.MongoLatency.WithLabelValues("invoices", "insert").Observe(time.Since(start).Seconds())
		r.metrics.MongoOperations.WithLabelValues("invoices", "insert").Inc()
	}()
	_, err := r.collection.InsertOne(ctx, invoice)
	return err
}

func (r *InvoiceRepository) GetByID(ctx context.Context, id string) (*entity.Invoice, error) {
	start := time.Now()
	defer func() {
		r.metrics.MongoLatency.WithLabelValues("invoices", "find_one").Observe(time.Since(start).Seconds())
		r.metrics.MongoOperations.WithLabelValues("invoices", "find_one").Inc()
	}()
	var inv entity.Invoice
	err := r.collection.FindOne(ctx, bson.M{"_id": id}).Decode(&inv)
	if err != nil {
		return nil, err
	}
	return &inv, nil
}

func (r *InvoiceRepository) ListByClient(ctx context.Context, clientID string, page, pageSize int) ([]*entity.Invoice, int64, error) {
	start := time.Now()
	defer func() {
		r.metrics.MongoLatency.WithLabelValues("invoices", "find").Observe(time.Since(start).Seconds())
		r.metrics.MongoOperations.WithLabelValues("invoices", "find").Inc()
	}()
	query := bson.M{"client_id": clientID}
	total, err := r.collection.CountDocuments(ctx, query)
	if err != nil {
		return nil, 0, err
	}
	if page < 1 {
		page = 1
	}
	if pageSize < 1 {
		pageSize = 20
	}
	opts := options.Find().SetSkip(int64((page - 1) * pageSize)).SetLimit(int64(pageSize)).SetSort(bson.M{"created_at": -1})
	cursor, err := r.collection.Find(ctx, query, opts)
	if err != nil {
		return nil, 0, err
	}
	defer cursor.Close(ctx)
	var invoices []*entity.Invoice
	if err := cursor.All(ctx, &invoices); err != nil {
		return nil, 0, err
	}
	return invoices, total, nil
}

func (r *InvoiceRepository) ListByProject(ctx context.Context, projectID string) ([]*entity.Invoice, error) {
	start := time.Now()
	defer func() {
		r.metrics.MongoLatency.WithLabelValues("invoices", "find").Observe(time.Since(start).Seconds())
		r.metrics.MongoOperations.WithLabelValues("invoices", "find").Inc()
	}()
	opts := options.Find().SetSort(bson.M{"created_at": -1})
	cursor, err := r.collection.Find(ctx, bson.M{"project_id": projectID}, opts)
	if err != nil {
		return nil, err
	}
	defer cursor.Close(ctx)
	var invoices []*entity.Invoice
	if err := cursor.All(ctx, &invoices); err != nil {
		return nil, err
	}
	return invoices, nil
}

func (r *InvoiceRepository) Update(ctx context.Context, invoice *entity.Invoice) error {
	start := time.Now()
	defer func() {
		r.metrics.MongoLatency.WithLabelValues("invoices", "update").Observe(time.Since(start).Seconds())
		r.metrics.MongoOperations.WithLabelValues("invoices", "update").Inc()
	}()
	invoice.UpdatedAt = time.Now()
	_, err := r.collection.ReplaceOne(ctx, bson.M{"_id": invoice.ID}, invoice)
	return err
}

var _ port.InvoiceRepository = (*InvoiceRepository)(nil)

File: apps/server/src/adapter/driven/mongodb/invoice-repository.ts
import { ObjectId } from "mongodb";
import type { Collection } from "mongodb";
import type { Invoice } from "../../../domain/entity/index.js";
import type { InvoiceRepository } from "../../../domain/port/index.js";
import type { MongoDBClient } from "./client.js";

interface InvoiceDoc {
  _id: ObjectId;
  project_id: string;
  client_id: string;
  invoice_number: string;
  status: string;
  line_items: Array<{
    id: string;
    description: string;
    quantity: number;
    unit_price: number;
    amount: number;
  }>;
  subtotal: number;
  tax: number;
  total: number;
  currency: string;
  due_date: Date;
  paid_at?: Date;
  payment_provider?: string;
  payment_reference?: string;
  notes?: string;
  created_at: Date;
  updated_at: Date;
}

/** Accept domain Invoice or BillingService shape (`items` / `total` / `paymentId`). */
function toDoc(invoice: Invoice | Record<string, any>): InvoiceDoc {
  const i = invoice as Record<string, any>;
  const lineItems =
    Array.isArray(i.lineItems) && i.lineItems.length > 0
      ? i.lineItems
      : Array.isArray(i.items)
        ? i.items.map((li: any) => ({
            id: li.id ?? new ObjectId().toHexString(),
            description: li.description,
            quantity: li.quantity,
            unitPrice: li.unitPrice ?? li.unit_price ?? 0,
            amount: li.amount ?? li.total ?? 0,
          }))
        : [];

  return {
    _id: new ObjectId(i.id),
    project_id: i.projectId,
    client_id: i.clientId,
    invoice_number: i.invoiceNumber,
    status: i.status,
    line_items: lineItems.map((li: any) => ({
      id: li.id ?? new ObjectId().toHexString(),
      description: li.description,
      quantity: li.quantity,
      unit_price: li.unitPrice ?? li.unit_price ?? 0,
      amount: li.amount ?? li.total ?? 0,
    })),
    subtotal: i.subtotal,
    tax: i.tax,
    total: i.total,
    currency: i.currency,
    due_date: i.dueDate,
    paid_at: i.paidAt,
    payment_provider: i.paymentProvider,
    payment_reference: i.paymentReference ?? i.paymentId,
    notes: i.notes,
    created_at: i.createdAt,
    updated_at: i.updatedAt,
  };
}

function fromDoc(d: InvoiceDoc): Invoice {
  return {
    id: d._id.toHexString(),
    projectId: d.project_id,
    clientId: d.client_id,
    invoiceNumber: d.invoice_number,
    status: d.status as Invoice["status"],
    lineItems: d.line_items.map((li) => ({
      id: li.id,
      description: li.description,
      quantity: li.quantity,
      unitPrice: li.unit_price,
      amount: li.amount,
    })),
    subtotal: d.subtotal,
    tax: d.tax,
    total: d.total,
    currency: d.currency,
    dueDate: d.due_date,
    paidAt: d.paid_at,
    paymentProvider: d.payment_provider as Invoice["paymentProvider"],
    paymentReference: d.payment_reference,
    notes: d.notes,
    createdAt: d.created_at,
    updatedAt: d.updated_at,
  };
}

export class MongoInvoiceRepository implements InvoiceRepository {
  private readonly col: Collection<InvoiceDoc>;

  constructor(client: MongoDBClient) {
    this.col = client.collection<InvoiceDoc>("invoices");
  }

  async findById(id: string): Promise<Invoice | null> {
    const doc = await this.col.findOne({ _id: new ObjectId(id) });
    return doc ? fromDoc(doc) : null;
  }

  async findByProjectId(projectId: string): Promise<Invoice[]> {
    const docs = await this.col
      .find({ project_id: projectId })
      .sort({ created_at: -1 })
      .toArray();
    return docs.map(fromDoc);
  }

  async findByClientId(clientId: string): Promise<Invoice[]> {
    const docs = await this.col
      .find({ client_id: clientId })
      .sort({ created_at: -1 })
      .toArray();
    return docs.map(fromDoc);
  }

  async listByClient(
    clientId: string,
    page: number = 1,
    pageSize: number = 20,
  ): Promise<{ invoices: Invoice[]; total: number }> {
    const query = { client_id: clientId };
    const total = await this.col.countDocuments(query);
    const docs = await this.col
      .find(query)
      .sort({ created_at: -1 })
      .skip((page - 1) * pageSize)
      .limit(pageSize)
      .toArray();
    return { invoices: docs.map(fromDoc), total };
  }

  async create(invoice: Invoice): Promise<Invoice> {
    await this.col.insertOne(toDoc(invoice));
    return invoice;
  }

  async update(idOrInvoice: string | Invoice, data?: Partial<Invoice>): Promise<Invoice> {
    if (typeof idOrInvoice === "string") {
      const existing = await this.findById(idOrInvoice);
      if (!existing) throw new Error(`Invoice not found: ${idOrInvoice}`);
      const merged = { ...existing, ...data, updatedAt: new Date() };
      const doc = toDoc(merged);
      await this.col.replaceOne({ _id: doc._id }, doc);
      return merged;
    }
    const doc = toDoc({ ...idOrInvoice, updatedAt: new Date() });
    await this.col.replaceOne({ _id: doc._id }, doc);
    return { ...idOrInvoice, updatedAt: doc.updated_at };
  }

  /**
   * Conditionally set status=paid only when the invoice is not already paid.
   * Returns the updated invoice, or null if missing / already paid.
   */
  async markPaidIfUnpaid(id: string, paymentId: string, paidAt: Date): Promise<Invoice | null> {
    const doc = await this.col.findOneAndUpdate(
      { _id: new ObjectId(id), status: { $ne: "paid" } },
      {
        $set: {
          status: "paid",
          paid_at: paidAt,
          payment_reference: paymentId,
          updated_at: paidAt,
        },
      },
      { returnDocument: "after" },
    );
    return doc ? fromDoc(doc) : null;
  }

  async delete(id: string): Promise<void> {
    await this.col.deleteOne({ _id: new ObjectId(id) });
  }
}
